Add a flag to wxUpdateUIEvent to tell if the item supports the check action

imciner2 opened the pull request:
Not all items support check being set in an UpdateUIEvent handler (a menu item will assert if you call `Check()` in the handler and the item doesn't support it), so it is nice to provide an API to find out if the item supports it. That way programs can make use of a single update UI event handler for their program. For instance, in the software I work on there is a single handler that gets condition functors bound to it at runtime before `Bind` is called, then the functors are evaluated at each event to tell it to enable/disable and check/uncheck the items. Currently, because the menu items assert when you call `Check()` in the event handler you have to use RTTI to figure out if the object being updated supports check, like the following code does: 
```
bool canCheck = true;

// wxMenuItems don't want to be checked unless they actually are checkable, so we have to check to
// see if they can be and can't just universally apply a check in this event.
if( auto menu = dynamic_cast<wxMenu*>( aEvent.GetEventObject() ) )
    canCheck = menu->FindItem( aEvent.GetId() )->IsCheckable();

if( canCheck )
    aEvent.Check( checkRes );
```

After this change, that code is much simpler (no need for the dynamic cast, and no need to find the item in the menu):
```
if( aEvent.IsCheckable() )
    aEvent.Check( checkRes );
```

I looked through the code that processes the `Check()` result in each of the locations that uses `wxUpdateUIEvent` and have replicated any condition checked as the setting for the `IsCheckable()` condition in the event, or if no condition is checked before it uses the result I just made it return true (there is one case inside `src/osx/menu_osx.cpp` where there was a null check for `item`, so I have made it set to true for the case when the item is null so it is backwards-compatible - but I don't know when that condition would ever be hit).

This also is something someone requested a while ago:
Fixes [#13369](https://trac.wxwidgets.org/ticket/13369)

imciner2 committed:
Add a flag to wxUpdateUIEvent to tell if the item supports the check action

Not all items support check being set in an UpdateUIEvent handler,
so it is nice to provide an API to find out if the item supports it.

Fixes #13369

vadz commented on include/wx/event.h:
```diff
@@ -2919,6 +2921,10 @@ class WXDLLIMPEXP_CORE wxUpdateUIEvent : public wxCommandEvent
 
+    // Set a flag saying if the object being updated supports the check action
+    bool IsCheckable() const { return m_isCheckable; }
+    void SetIsCheckable(bool aCheckable) { m_isCheckable = aCheckable; }
```
We don't use `aFoo` convention anywhere else, so I'd rather do
```suggestion
    void SetIsCheckable(bool checkable) { m_isCheckable = checkable; }
```

vadz commented on include/wx/event.h:
```diff
@@ -2944,6 +2950,7 @@ class WXDLLIMPEXP_CORE wxUpdateUIEvent : public wxCommandEvent
     virtual wxEvent *Clone() const wxOVERRIDE { return new wxUpdateUIEvent(*this); }
 
 protected:
+    bool          m_isCheckable;
```
For consistency with the initialization order I'd put this after `m_setChecked`.

vadz commented on src/osx/menu_osx.cpp:
```diff
@@ -324,6 +324,11 @@ bool wxMenu::HandleCommandUpdateStatus( wxMenuItem* item, wxWindow* senderWindow
     wxUpdateUIEvent event(menuid);
     event.SetEventObject( this );
 
+    if ( item )
+        event.SetIsCheckable(item->IsCheckable());
+    else
+        event.SetIsCheckable(true);
```
Why is the event checkable when there is no item? Shouldn't this rather be
```suggestion
    event.SetIsCheckable(item && item->IsCheckable());
```
?

vadz reviewed (commented):
Thanks, no real comments, I agree that this may be useful.

From compatibility point of view, however, I think we should set `m_isCheckable = true` by default and provide `void DisallowCheck()` setting it to `false`. The risk of breaking some code creating its own `wxUpdateUIEvent`s seems slight, but it's still non zero and it looks like it wouldn't cost us anything to do it like this (in fact it would even avoid having to change ribbon code).

Could you please update this PR to do it like this or do you see any problems with this approach?

imciner2 committed:
Make IsCheckable true by default and switch to explicitly disallowing it

Also add a new accessor to wxAuiToolBarItem to make it easier to
determine when a tool can be checked.

imciner2 committed:
Use new CanBeToggled function inside wxAuiToolBar where appropriate

vadz reviewed (approved):
Oops, sorry for missing the update to this one.

Looks perfect to me, I'll just add one small line to the docs and merge it in a moment.

Thanks!
